wolfcrypt/src/port/aria/aria-cryptocb.c

```diff
@@ -0,0 +1,528 @@
+        return 0;
+    }
+
+    /* return 0 on success or BAD_FUNC_ARG on failure */
+    int wc_AriaShaFinal(MC_HSESSION hSession, byte* out, word32* len)
+    {
+        MC_RV rv = MC_OK;
+        WOLFSSL_ENTER("AriaShaFinal");
+
+        if (out == NULL || len == NULL) {
+            return BAD_FUNC_ARG;
+        }
+
+        /* Do an extra DigestUpdate noop just in case it is never explicitly called. */
+        if (rv == MC_OK) rv = MC_DigestUpdate(hSession, NULL, 0);
+
+        if (rv == MC_OK) rv = MC_DigestFinal(hSession, out, len);
+
+        if (rv != MC_OK) {
+            WOLFSSL_MSG(MC_GetErrorString(rv));
+            return BAD_FUNC_ARG;
+        }
+        /* WOLFSSL_MSG_EX("Digest len: %d", *len); */
+        return 0;
+    }
+
+    /* return 0 on success or BAD_STATE_E on failure */
+    int wc_AriaFree(MC_HSESSION* hSession, MC_HOBJECT *obj1)
+    {
+        MC_RV rv = MC_OK;
+        WOLFSSL_ENTER("AriaFree");
+
+        if (hSession == NULL && obj1 != NULL) {
+            return BAD_FUNC_ARG;
+        }
+
+        if (obj1 != NULL) {
+            if (rv == MC_OK) rv = MC_DestroyObject(*hSession, *obj1);
+            if (rv == MC_OK) *obj1 = NULL;
+        }
+
+        if (hSession != NULL) {
+            if (rv == MC_OK) rv = MC_CloseSession(*hSession);
+            if (rv == MC_OK) *hSession = NULL;
+        }
+
+        if (rv != MC_OK) {
+            WOLFSSL_MSG(MC_GetErrorString(rv));
+            return BAD_STATE_E;
+        }
+        return 0;
+    }
+
+    int wc_AriaSign(byte* in, word32 inSz, byte* out, word32* outSz, ecc_key* key)
+    {
+        MC_HOBJECT hPrikey = 0;
+        MC_HSESSION hSession = 0;
+
+        const ecc_set_type* dp;
+        MC_RV rv = MC_OK;
+
+        MC_APIMODE gApimode = MC_MODE_KCMV;
+        MC_ALGORITHM mcAlg = {MC_ALGID_NONE, NULL, 0};
+        byte keyAsn1[128];
+        word32 keyAsn1Sz=sizeof(keyAsn1);
+
+        WOLFSSL_ENTER("AriaSign");
+
+        if (in == NULL || out == NULL || outSz == NULL || key == NULL) {
+            return BAD_FUNC_ARG;
+        }
+
+        if (rv == MC_OK) rv = wc_AriaInit();
+
+        if (rv == MC_OK) rv = MC_OpenSession(&hSession);
+
+        if (rv == MC_OK) rv = MC_SetApiMode(hSession, gApimode);
+
+        if (rv == MC_OK) {
+            int ret = wc_EccPrivateKeyToDerNoCurve(key,keyAsn1,keyAsn1Sz);
+            if (ret < 0) { rv = ret; }
+            else { keyAsn1Sz = ret; }
+        }
+
+        WOLFSSL_MSG_EX("AriaSign key(%d):",keyAsn1Sz);
+        WOLFSSL_BUFFER(keyAsn1,keyAsn1Sz);
+
+        WOLFSSL_MSG_EX("AriaSign rv=%d",rv);
+
+        if (key->dp != NULL) {
+            dp = key->dp;
+        }
+        else {
+            dp = wc_ecc_get_curve_params(key->idx);
+        }
+
+        if (dp->id == ECC_SECP256R1) {
+            mcAlg.mcAlgId = MC_ALGID_SHA256WithECDSA_P256_r1;
+        } else if (dp->id == ECC_SECP224R1) {
+            mcAlg.mcAlgId = MC_ALGID_SHA256WithECDSA_P224_12;
+        } else {
+            return BAD_FUNC_ARG;
```

Comment:
Is cleanup handled correctly in this case?

Reply:
Good catch.